bionic/core/flow_execution.py

```diff
@@ -155,8 +165,66 @@ def _process_active_entry(self, entry):
             and not entry.state.task.is_simple_lookup
         )
         if can_compute_remotely:
-            new_state_for_subprocess = entry.state.strip_state_for_subprocess()
-            # We want serial execution in the subprocesses.
+            # When we run an entry remotely, we also need to run all of its immediate
+            # non-persistable ancestors, since their values can't be shared between
+            # processes. Those ancestors may have follow-up tasks, so we'll need to
+            # run those too; and if those follow-ups are persistable (but not actually
+            # persisted yet), their values will be usable everywhere, so we'll want to
+            # track the fact that they're running. So really we're going to run a
+            # collection of persistable entries (including the original entry), plus
+            # their immediate non-persistable ancestors.
+            remote_subgraph = RemoteSubgraph(entry.state, self._bootstrap)
+            target_entries = [
+                self._get_or_create_entry_for_state(persistable_subgraph_state)
+                for persistable_subgraph_state in remote_subgraph.persistable_but_not_persisted_states
+            ]
+            assert entry in target_entries
+
+            # We want all of our entries to be initialized before we try running
+            # them remotely. We'll just require this of the persistable ones, since
+            # all the other ones are ancestors of them.
+            for target_entry in target_entries:
+                if target_entry == entry:
+                    assert target_entry.level >= EntryLevel.INITIALIZED
+                else:
+                    self._add_requirement(
+                        src_entry=entry,
+                        dst_entry=target_entry,
+                        level=EntryLevel.INITIALIZED,
+                    )
+            if not entry.all_outgoing_reqs_are_met:
+                self._mark_entry_blocked(entry)
+                return
+
+            for target_entry in target_entries:
+                # Now we'll also require that each entry be cached, which guarantees
```

Comment:
This is very neat (and a little tricky)! I understand now why we need it, but I'm sure I'll be a little confused when I see this code later down the road. Can you also add a line or two about how it can be `COMPLETED` without this requirement (which IIUC, is because the follow up dependency might not have anyone consuming it yet)?

Reply:
Yep, adding more discussion here makes sense.

Also, I'm really hoping we can simplify away a lot of this by introducing file descriptors.